core/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -348,21 +876,55 @@ public void visit(Visitor visitor)
     visitor.visit(this);
   }
 
+  @Override
+  public Expr visit(Shuttle shuttle)
+  {
+    Expr newLeft = left.visit(shuttle);
+    Expr newRight = right.visit(shuttle);
+    if (left != newLeft || right != newRight) {
+      return shuttle.visit(copy(newLeft, newRight));
+    }
+    return shuttle.visit(this);
+  }
+
   @Override
   public String toString()
   {
-    return "(" + op + " " + left + " " + right + ")";
+    return StringUtils.format("(%s %s %s)", op, left, right);
+  }
+
+  protected abstract BinaryOpExprBase copy(Expr left, Expr right);
+
+  @Override
+  public BindingDetails analyzeInputs()
+  {
+    // currently all binary operators operate on scalar inputs
+    final Set<String> scalars = new HashSet<>();
+    final String leftIdentifer = left.getIdentifierIfIdentifier();
+    final String rightIdentifier = right.getIdentifierIfIdentifier();
+    if (leftIdentifer != null) {
+      scalars.add(leftIdentifer);
+    }
+    if (rightIdentifier != null) {
+      scalars.add(rightIdentifier);
+    }
+    return left.analyzeInputs()
+               .merge(right.analyzeInputs())
+               .mergeWithScalars(scalars);
```

Comment:
Ah, so the operators and functions have to add context of how they are using `IdentifierExpr` as part of their work for `analyzeInputs`. This is currently done in a shallow manner, in other words, the operators and functions only add direct `IdentifierExpr` children to 'scalar' or 'array' collections based on the type of functions. `IdentifierExpr` themselves do not know their context, so only add themselves to the set of 'free' variables.

In case an example makes this makes more sense, imagine an expression: `x + (y - z)` which is parsed into `Expr` as the tree:

```
               BinPlusExpr
               /         \
   IdentifierExpr(x)  BinMinusExpr
                     /            \
          IdentifierExpr(y)   IdentifierExpr(z)
```
or
```
   +
 /   \
x     -
    /   \
   y     z
```
for short.

working this out:
```
analyzeInputs(+) = analyzeInputs(x) + analyzeInputs(-) + "set of direct IdentifierExpr children of +"
analyzeInputs(+) = (analyzeInputs(x)) + (analyzeInputs(y) + analyzeInputs(z) + "set of direct IdentifierExpr children of -") + ("set of direct IdentifierExpr children of +")
analyzeInputs(+) = ([free:[x], scalar:[], array:[]]) + ([free:[y], scalar:[], array:[]] + [free:[z], scalar:[], array:[]] + [scalar:[y, z]]) + ([scalar:[x]])
analyzeInputs(+) = [free:[x, y, z], scalar:[x, y, z], array:[]]
```

Hopefully that didn't make it more confusing!

This approach definitely has some shortcomings, such as if a function works with both array and scalar inputs such as the accumulator input of the 'fold' functions, but luckily there is very little overlap between array and scalar functions right now. I imagine that this `analyzeInputs` machinery will be largely reworked or replaced or scrapped in the future when I try to add full type inference necessary to allow us to fully optimize expressions to use primitive types where possible, but for now it's doing ok enough to usually give me enough context to determine when an input column is expected to be an array, if an array input is _not_ being used as an array, and if any contradictory usages of an input column as an array and a scalar are detected.

Reply:
thanks for explanation